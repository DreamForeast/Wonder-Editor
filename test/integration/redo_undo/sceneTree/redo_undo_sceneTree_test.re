open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("redo_undo: sceneTree", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => sandbox := createSandbox());
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
    describe("get scene tree from engine", () => {
      let _simulateTwiceDragEvent = () => {
        let firstCameraDomIndex =
          SceneTreeNodeDomTool.OperateTwoLayer.getFirstCameraDomIndex();
        let firstCubeDomIndex =
          SceneTreeNodeDomTool.OperateTwoLayer.getFirstCubeDomIndex();
        let secondCubeDomIndex =
          SceneTreeNodeDomTool.OperateTwoLayer.getSecondCubeDomIndex();

        let component =
          BuildComponentTool.buildSceneTree(
            TestTool.buildAppStateSceneGraphFromEngine(),
          );
        BaseEventTool.triggerComponentEvent(
          component,
          SceneTreeEventTool.triggerDragStart(secondCubeDomIndex),
        );
        BaseEventTool.triggerComponentEvent(
          component,
          SceneTreeEventTool.triggerDragEnter(firstCameraDomIndex),
        );
        BaseEventTool.triggerComponentEvent(
          component,
          SceneTreeEventTool.triggerDragDrop(firstCameraDomIndex),
        );

        let component2 =
          BuildComponentTool.buildSceneTree(
            TestTool.buildAppStateSceneGraphFromEngine(),
          );
        BaseEventTool.triggerComponentEvent(
          component2,
          SceneTreeEventTool.triggerDragStart(firstCubeDomIndex),
        );
        BaseEventTool.triggerComponentEvent(
          component2,
          SceneTreeEventTool.triggerDragEnter(firstCameraDomIndex),
        );
        BaseEventTool.triggerComponentEvent(
          component2,
          SceneTreeEventTool.triggerDragDrop(firstCameraDomIndex),
        );
      };
      beforeEach(() => {
        TestTool.closeContractCheck();
        MainEditorSceneTool.initStateAndGl(~sandbox, ());
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
        );
        StateHistoryToolEditor.clearAllState();
      });
      afterEach(() => TestTool.openContractCheck());
      describe("test undo operate", () => {
        test("test not undo", () => {
          _simulateTwiceDragEvent();

          BuildComponentTool.buildSceneTree(
            TestTool.buildAppStateSceneGraphFromEngine(),
          )
          |> ReactTestTool.createSnapshotAndMatch;
        });
        describe("test undo one step", () =>
          test("step which from second to first", () => {
            _simulateTwiceDragEvent();

            StateHistoryToolEditor.undo();

            BuildComponentTool.buildSceneTree(
              TestTool.buildAppStateSceneGraphFromEngine(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );
        describe("test undo two step", () =>
          test("step which from second to zero", () => {
            _simulateTwiceDragEvent();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();
            BuildComponentTool.buildSceneTree(
              TestTool.buildAppStateSceneGraphFromEngine(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );
        describe("test undo three step", () =>
          test("if current step is zero, undo should do nothing", () => {
            _simulateTwiceDragEvent();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();
            BuildComponentTool.buildSceneTree(
              TestTool.buildAppStateSceneGraphFromEngine(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );
      });
      describe("test redo operate", () => {
        describe("test redo one step", () => {
          test("if not exec undo, redo one step should do nothing", () => {
            _simulateTwiceDragEvent();
            StateHistoryToolEditor.redo();
            BuildComponentTool.buildSceneTree(
              TestTool.buildAppStateSceneGraphFromEngine(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
          test(
            "undo step which from second to zero, redo step which from zero to first",
            () => {
            _simulateTwiceDragEvent();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.redo();
            BuildComponentTool.buildSceneTree(
              TestTool.buildAppStateSceneGraphFromEngine(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          });
        });
        describe("test redo two step", () =>
          test(
            "undo step which from second to zero, redo step which from zero to second",
            () => {
            _simulateTwiceDragEvent();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.redo();
            StateHistoryToolEditor.redo();
            BuildComponentTool.buildSceneTree(
              TestTool.buildAppStateSceneGraphFromEngine(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );
        describe("test redo three step", () =>
          test("test if current step is last step, redo should do nothing", () => {
            _simulateTwiceDragEvent();

            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.undo();
            StateHistoryToolEditor.redo();
            StateHistoryToolEditor.redo();
            StateHistoryToolEditor.redo();
            BuildComponentTool.buildSceneTree(
              TestTool.buildAppStateSceneGraphFromEngine(),
            )
            |> ReactTestTool.createSnapshotAndMatch;
          })
        );
      });
    });
    describe("fix bug", () => {
      let execSetCurrentSceneTreeNodeWork = () => {
        let component =
          BuildComponentTool.buildSceneTree(
            TestTool.buildAppStateSceneGraphFromEngine(),
          );
        let secondCubeDomIndex =
          SceneTreeNodeDomTool.OperateTwoLayer.getSecondCubeDomIndex();

        BaseEventTool.triggerComponentEvent(
          component,
          SceneTreeEventTool.triggerClickEvent(secondCubeDomIndex),
        );
      };
      let execChangeMaterialColorWork = (currentGameObjectMaterial, newColor) =>
        MaterialEventTool.triggerChangeColor(
          currentGameObjectMaterial,
          newColor,
        );

      let execChangeTransformWork = () => {
        let currentGameObjectTransform =
          GameObjectTool.getCurrentSceneTreeNodeTransform();
        let transformComponent =
          BuildComponentTool.buildMainEditorTransformComponent(
            TestTool.buildEmptyAppState(),
            currentGameObjectTransform,
          );
        BaseEventTool.triggerComponentEvent(
          transformComponent,
          TransformEventTool.triggerChangeXEvent("11.25"),
        );
        BaseEventTool.triggerComponentEvent(
          transformComponent,
          TransformEventTool.triggerBlurXEvent("11.25"),
        );
      };

      beforeEach(() => {
        MainEditorSceneTool.initStateAndGl(~sandbox, ());
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
        );
      });
      test(
        "the workflow: click treeNote set currentSceneTreeNode -> change material color -> change transform x value -> undo, engineState is error",
        () => {
          let currentGameObjectMaterial =
            GameObjectTool.getCurrentGameObjectMaterial();
          let newColor = {
            "hex": "#7df1e8",
            "rgb": {
              "r": 125,
              "g": 241,
              "b": 232,
            },
          };

          execChangeMaterialColorWork(currentGameObjectMaterial, newColor);
          execChangeTransformWork();
          StateHistoryToolEditor.undo();

          BasicMaterialEngineService.getColor(currentGameObjectMaterial)
          |> StateLogicService.getEngineStateToGetData
          |> Color.getHexString
          |> expect == newColor##hex;
        },
      );
    });
  });