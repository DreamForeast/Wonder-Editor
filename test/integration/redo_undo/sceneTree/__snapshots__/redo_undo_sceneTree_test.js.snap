// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redo_undo: sceneTree get scene tree from engine test redo operate test redo one step if not exec undo, redo one step should do nothing 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-drag-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        camera
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "border": "1px solid red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "background": "red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
    </ul>
    <div
      className="wonder-disable-drag"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "backgroundColor": "#c0c0c0",
        }
      }
    />
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test redo operate test redo one step undo step which from second to zero, redo step which from zero to first 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-drag-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        camera
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "border": "1px solid red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
    </ul>
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "background": "red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        cube
      </li>
    </ul>
    <div
      className="wonder-disable-drag"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "backgroundColor": "#c0c0c0",
        }
      }
    />
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test redo operate test redo three step test if current step is last step, redo should do nothing 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-drag-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        camera
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "border": "1px solid red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "background": "red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
    </ul>
    <div
      className="wonder-disable-drag"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "backgroundColor": "#c0c0c0",
        }
      }
    />
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test redo operate test redo two step undo step which from second to zero, redo step which from zero to second 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-drag-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        camera
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "border": "1px solid red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "background": "red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
    </ul>
    <div
      className="wonder-disable-drag"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "backgroundColor": "#c0c0c0",
        }
      }
    />
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test undo operate test not undo 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-drag-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        camera
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "border": "1px solid red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "background": "red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
    </ul>
    <div
      className="wonder-disable-drag"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "backgroundColor": "#c0c0c0",
        }
      }
    />
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test undo operate test undo one step step which from second to first 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-drag-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        camera
      </li>
      <ul
        className="wonder-tree-node"
        draggable={true}
        onDragEnd={[Function]}
        onDragStart={[Function]}
      >
        <li
          onClick={[Function]}
          style={
            Object {
              "border": "1px solid red",
            }
          }
        >
          <div
            className="item-ground"
            draggable={true}
            onDragEnter={[Function]}
            onDragLeave={[Function]}
            onDragOver={[Function]}
            onDrop={[Function]}
          />
          cube
        </li>
      </ul>
    </ul>
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "background": "red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        cube
      </li>
    </ul>
    <div
      className="wonder-disable-drag"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "backgroundColor": "#c0c0c0",
        }
      }
    />
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test undo operate test undo three step if current step is zero, undo should do nothing 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-drag-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        camera
      </li>
    </ul>
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "background": "red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        cube
      </li>
    </ul>
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        cube
      </li>
    </ul>
    <div
      className="wonder-disable-drag"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "backgroundColor": "#c0c0c0",
        }
      }
    />
  </article>
</article>
`;

exports[`redo_undo: sceneTree get scene tree from engine test undo operate test undo two step step which from second to zero 1`] = `
<article
  className="wonder-sceneTree-component"
>
  <article
    className="wonder-drag-tree"
  >
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        camera
      </li>
    </ul>
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "background": "red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        cube
      </li>
    </ul>
    <ul
      className="wonder-tree-node"
      draggable={true}
      onDragEnd={[Function]}
      onDragStart={[Function]}
    >
      <li
        onClick={[Function]}
        style={
          Object {
            "border": "1px solid red",
          }
        }
      >
        <div
          className="item-ground"
          draggable={true}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
        />
        cube
      </li>
    </ul>
    <div
      className="wonder-disable-drag"
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDrop={[Function]}
      style={
        Object {
          "backgroundColor": "#c0c0c0",
        }
      }
    />
  </article>
</article>
`;
